dbms/src/Storages/DeltaMerge/VersionChain/DMFileHandleIndex.h

```diff
@@ -0,0 +1,241 @@
+        , clipped_pack_range(getPackRange(dm_context_))
+        , clipped_pack_index(loadPackIndex(dm_context_))
+        , clipped_pack_offsets(loadPackOffsets())
+        , clipped_handle_packs(clipped_pack_range.count())
+        , clipped_need_read_packs(clipped_pack_range.count(), 1) // read all packs by default
+    {
+        RUNTIME_CHECK(
+            clipped_pack_index.size() == clipped_pack_range.count(),
+            clipped_pack_index.size(),
+            clipped_pack_range.count());
+        RUNTIME_CHECK(
+            clipped_pack_offsets.size() == clipped_pack_range.count(),
+            clipped_pack_offsets.size(),
+            clipped_pack_range.count());
+    }
+
+    // Get the base(oldest) version of the given handle.
+    // 1. Get the pack may contain the given handle.
+    // 2. Get the base version of the given handle in the given pack.
+    std::optional<RowID> getBaseVersion(const DMContext & dm_context, HandleRefType h)
+    {
+        auto clipped_pack_id = getClippedPackId(h);
+        if (!clipped_pack_id)
+            return {};
+        auto row_id = getBaseVersion(dm_context, h, *clipped_pack_id);
+        if (!row_id)
+            return {};
+        return start_row_id + *row_id;
+    }
+
+    // Use to pre-calculate which packs to read.
+    template <typename Iterator>
+    void calculateReadPacks(Iterator begin, Iterator end)
+    {
+        std::vector<UInt8> calc_read_packs(clipped_pack_range.count(), 0);
+        UInt32 calc_read_count = 0;
+        for (auto itr = begin; itr != end; ++itr)
+        {
+            auto clipped_pack_id = getClippedPackId(*itr);
+            if (!clipped_pack_id || calc_read_packs[*clipped_pack_id])
+                continue;
+
+            calc_read_packs[*clipped_pack_id] = 1;
+            ++calc_read_count;
+
+            // Read too many packs, read all for simplicity.
+            // TODO: optimize the storage format of handle column to improve the performance.
+            if (calc_read_count * 4 >= clipped_pack_range.count())
+                return; // return, instead of break, because `clipped_need_read_packs` is read all by default.
+        }
+        clipped_need_read_packs = std::move(calc_read_packs);
+    }
+
+    // Release handle column data to save memory.
+    void cleanHandleColumn()
+    {
+        std::fill(clipped_handle_packs.begin(), clipped_handle_packs.end(), nullptr);
+        clipped_need_read_packs = std::vector<UInt8>(clipped_pack_range.count(), 1);
+    }
+
+private:
+    // Find the pack may contian the given handle.
+    std::optional<UInt32> getClippedPackId(HandleRefType h) const
+    {
+        if (unlikely(rowkey_range && !inRowKeyRange(*rowkey_range, h)))
+            return {};
+
+        auto itr = std::lower_bound(clipped_pack_index.begin(), clipped_pack_index.end(), h);
+        if (itr == clipped_pack_index.end())
+            return {};
+        return itr - clipped_pack_index.begin();
+    }
```

Comment:
```suggestion
    // Find the pack may contian the given handle.
    UInt32 getClippedPackId(HandleRefType h) const
    {
        if (unlikely(rowkey_range && !inRowKeyRange(*rowkey_range, h)))
            return clipped_pack_index.size();

        auto itr = std::lower_bound(clipped_pack_index.begin(), clipped_pack_index.end(), h);
        return std::distance(clipped_pack_index.begin(), itr);
    }
```

This can reduece some if-branch. And check `getClippedPackId(h) < clipped_pack_index.size()`